### ISSUE TITLE
AMF bundling for OSGi

### ISSUE THREAD

**Issue by jgsuess** [aml-org/amf#1228]
Your issue may already be reported! Please search on Github issues before creating one.

* **I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] question
  - [ ] other

* **What is the current behavior?**
[The `client` packages and potentially others are split over several jars](https://nirmalsasidharan.wordpress.com/2010/09/02/split-packages-an-osgi-nightmare/). This makes direct reuse of the archives as bundles very challenging and blocks fluent use in OSGi and Eclipse as a result.

* **If the current behavior is a bug, please provide the minimal steps to reproduce with AMF code. Please don't upload AMF code mixed with yours.**

* **What is the expected behavior?**

Packages should not be split, produce one assembled archive or OSGi bundle.

* **What is the motivation/use case for changing the behavior? (for feature requests)**

* **Please tell us about your environment:**
  
  - AMF Version: all
  - AMF Distribution: JVM
  - Operating System: all
  - Link to your project:

* **Other information** (e.g. detailed explanation, stack traces, related issues, suggestions how to fix, links for us to have context, eg. StackOverflow, Gitter, etc)

**Comment by AgustinBettati** [comment]
Hello @jgsuess!

Could you provide a specific version where you encountered this issue? 
From my understanding this problem occurs when there are separate libraries that make use of the same package name. For the case of AMF versions 5.x.x we can ensure that this is not posible as we have normalized our package structure, where the packages contained in each library have a unique prefix.

**Comment by jgsuess** [comment]
I am using amf-core_2.12-4.1.20, which is consistent with your specification. However, this version is used transitively used from webapi-parser-0.5.0.jar: com.github.amlorg » amf-webapi_2.12 | 4.0.3

Hence, I cannot profit from the OSGi bundling.

From an end-users perspective, the AMF-AML is valuable for working with RAML. No-one will go out and write a chocolate factory control modelling software using AMF. The entry-point to services hence is the parser bundle for RAML. That should be up-to-date in the public space.

**Comment by AgustinBettati** [comment]
Regarding **webapi-parser**: Since the release of AMF 5.x.x, this repository has not been maintained and is replaced in favour of the new interfaces provided directly by AMF (I am currently working on making this clear in the webapi-parser repository). All of the functionality related to API Specs can be invoked in a very simple manner.

On the other hand, I may have misunderstood your underlying problem. Is it related to the fact that different jars use same package names, or simply because you cannot obtain a single bundle?

**Comment by jgsuess** [comment]
If **webapi-parser** is deprecated after **[RAML Java Parser](https://github.com/raml-org/raml-java-parser)** is deprecated, then pointing out what is current would be valuable. A standard like RAML without a TCK and accessible reference implementation is not very useful.

As to my problem, I am using [bnd](https://bnd.bndtools.org/) with either [maven P2](https://github.com/reficio/p2-maven-plugin) or Tycho P2 platform targets. All use bnd on a single bundle under the hood. BND adds the OSGi metadata transitively. Webapi  parser cannot be used this way, because AMF contains identical package names in different jars. This is known as split packages, and is an anti-pattern for OSGi.

I had to manually download the transitive dependency tree and then bundle it all inside a single OSGi bundle on the internal classpath. 

* Ugly, as it leads to huge memory wastage as I have most of the bundles in the system already. 
* Slow, because the bundle is enormous as a download. I cannot use RAML on mobile devices and am building on OAS for that as a result. That is much smaller, due to OSGi sharing. 
* A maintenance pain, as the resources are bundled and for transitive version updates I need to rerun the build, rather than update the bundle in the OSGi runtime.  
* A security loophole, due to same update issues, think log4j.

All in all, I am telling clients not to use RAML for their component services as the use under OSGi is not transparent or blocked by setup. I also advise against use in mobile as the footprint doubles with the current setup.

**Comment by AgustinBettati** [comment]
Thank you for all the insight and detail.

RAML does in fact have a [TCK](https://github.com/raml-org/raml-tck) where you will also find reference to implementations. We will look into giving the AMF implementation more visibility on this end.

Given that the problem originates from identical package names in different jars, I advise you migrate from webapi-parser to AMF 5 directly. You will be using the up-to-date implementation for RAML processing (you will also find that we support oas and async), and this will also ensure no clashes in package names.

Closing this issue but feel free to comment if you have any doubts.

**Comment by jgsuess** [comment]
Could you please point me to the equivalent information for the parser invocation. Saying that I should 'migrate from `webapi-parser` to AMF 5 directly' is a bit like saying that I should 'look at the internet to read the documentation'. In other words, this is not a reasonably actionable, and hence not a useful, suggestion. Can you please provide either specific instructions or an example as to *how* parsing and serialisation are performed with this solution, or test code that exercises these functionalities. 

If you prefer, I may open another ticket with subject 'document how to use AMF 5 parser and serializer', given you would prefer to keep this one closed.

Thank you.

**Comment by jgsuess** [comment]
Self-note: It seems that the [generalized parsing instructions are here](https://a.ml/docs/amf/using-amf/amf_parsing#api-contract-model-parsing).

**Comment by jgsuess** [comment]
I have bundled AMF 5 up, and it does not declare its dynamic dependencies hence uses of the parser still fail. AMF5 cannot be used in OSGi.

@AgustinBettati please reopen.

```
Caused by: java.lang.ClassNotFoundException: org.everit.json.schema.internal.URIV4FormatValidator cannot be found by com.github.amlorg.amf-shapes_2.12_5.0.0
	at org.eclipse.osgi.internal.loader.BundleLoader.generateException(BundleLoader.java:516)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass0(BundleLoader.java:511)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:403)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 30 more
```

**Comment by jgsuess** [comment]
This either requires dynamic import (not recommended) or import specification.

http://keheliya.blogspot.com/2013/02/use-of-dynamicimport-package-in-osgi.html

**Comment by AgustinBettati** [comment]
Yes that documentation is accurate, happy to see you found it.

Reopening the issue as it is relevant to AMF 5. From the error I see there is class from one of our external dependencies that is not resolved, is this due to package naming from their side? If this is the case there is not much we can do to change that. Are there any other alternatives we have from our side to fix the OSGi bundling problem?

**Comment by jgsuess** [comment]
The cause is reflective loading performed by AMF 5 dependencies. AMF 5 can solve this by providing an OSGi manifest that lists all packages that may be dynamically loaded. This is something that a bundle analyzer like bnd cannot do, as the information is not explicit. Alternatively [Dynamic Imports](http://docs.osgi.org/specification/osgi.core/7.0.0/framework.module.html#framework.module.dynamicimportpackage) can be declared using wildcards. This is a security loophole to varying degrees, though. Without someone constructing a test in OSGi using bnd, I think AMF 5 will not become usable in OSGi / Eclipse in bundle form.
The other approach that can be taken is to create a project with all AMF5 libraries bundled and then try to extract them. This will gradually identify the issues. I might do that in the long run to reduce the situation I have.

**Comment by AgustinBettati** [comment]
Hello @jgsuess,

We have analyzed this issue with the team and have come to the conclusion that given our available resources and the limited value it brings to the project we will not be implementing or analyzing a related solution. I advise you generate a compliant bundle in a separate project with AMF dependencies, or feel free to contribute to the project if you find a solution from our end.

**Comment by jgsuess** [comment]
Thanks for performing the analysis. I understand it is challenging to make changes once dependencies have become tangled.
